File: TPs/Q01/palindromo.java
/*
 * Gabriel Xavier Borges - 805347 - Q01
 */

class palindromo
{   
    /*
     * Funcao para verificar condicao de parada
     * @param string
     * @return boolean
    */
    public static boolean isEnd (String str)
    {
        if(str.length() == 3 && str.charAt(0) == 'F' && str.charAt(1) == 'I' && str.charAt(2) == 'M')
        {
            return true;
        }
        return false;
    }

    /*
     * Funcao para verificar se e palindromo
     * @param string
     * @return boolean 
    */
    public static boolean isPalindromo (String str)
    {
        int j = str.length() - 1;
        boolean resp = true;
        for(int i = 0; i < str.length(); i++)
        {
            //System.out.println(""+str.charAt(i) + str.charAt(j));
            if(str.charAt(i) != str.charAt(j))
            {
                resp = false;
            }
            j--;
        }
        return resp;
    }

    /**
     * Funcao principal
     * @param agrs
    */
    public static void main (String[] agrs)
    {   
        MyIO.setCharset("US-ASCII"); //setCharset pois o padrao nao identifica alguns caracteres

        String str;
        str = MyIO.readLine("");

        do
        {
            if(isPalindromo(str))
            {
                System.out.println("SIM");
            }
            else
            {
                System.out.println("NAO");
            }
            str = MyIO.readLine("");
        }while(!isEnd(str));
    }
}

File: TPs/Q02/palindromo.c
/**
 * Gabriel Xavier Borges - 805347 - 2
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

/**
 * Funcao para verificar se e palindromo
 * @param string
 * @return bool
 */
bool isPalindromo(char *str)
{
    bool r = true;
    int j = strlen(str) - 1;
    for (int i = 0; i < strlen(str) / 2; i++)
    {
        if (str[i] != str[j])
        {
            r = false;
            i = strlen(str) / 2;
        }
        j--;
    }
    return r;
}//Existem erros de charset

/**
 * Funcao para ler linha
 * @return linha
 */
char *readline()
{
    char *str = (char *)calloc(1000, sizeof(char));
    if (str != NULL)
    {
        fgets(str, 1000, stdin);
        int i = strlen(str);
        if (i > 0 && str[i - 1] == '\n')
        {
            str[i - 1] = '\0';
        }
    }
    return str;
}

/**
 * Funcao principal
 */
int main(int agrc, char *agrv[])
{
    char *str = NULL;
    str = readline();
    do
    {
        if (isPalindromo(str) == true)
        {
            printf("SIM\n");
        }
        else
        {
            printf("NAO\n");
        }
        str = readline();
    } while (strcmp(str, "FIM\0") != 0);

    return 0;
}